crates/libs/windows/src/core/unknown.rs

```diff
@@ -61,15 +61,28 @@ impl core::fmt::Debug for IUnknown {
 pub trait IUnknownImpl {
     fn get_impl(&mut self) -> RawPtr;
 
-    fn QueryInterface(&mut self, iid: &GUID, interface: *mut RawPtr) -> HRESULT;
-    fn AddRef(&mut self) -> u32;
-    fn Release(&mut self) -> u32;
+    /// The classic `QueryInterface` method from COM.
+    ///
+    /// # Safety
+    ///
+    /// This function is safe to call as long as the interface pointer is non-null and valid for writes
+    /// of an interface pointer.
+    unsafe fn QueryInterface(&self, iid: &GUID, interface: *mut *const core::ffi::c_void) -> HRESULT;
```

Comment:
You could use [`NonNull`](https://doc.rust-lang.org/std/ptr/struct.NonNull.html) to encode the non-null safety requirement here if you wanted.

Reply:
I consider that, but I'm not sure it's worth it since `NonNull` is a bit awkward to work with, and the user still has to verify the pointer is valid for writing an interface pointer to. Since `NULL` is not valid for writes, the safety check should already tip the caller to check for non-null.